rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/responder.hpp

```diff
@@ -164,6 +164,49 @@ class Responder
     return estr;
   }
 
+  void deinit() noexcept
+  {
+    DDS::ReturnCode_t status;
+    if (response_datawriter_) {
+      status = response_publisher_->delete_datawriter(response_datawriter_);
+      if (nullptr != impl::check_delete_datawriter(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_datawriter(status));
+      }
+    }
+    if (response_topic_) {
+      status = participant_->delete_topic(response_topic_);
+      if (nullptr != impl::check_delete_topic(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_topic(status));
+      }
+    }
+    if (response_publisher_) {
+      status = participant_->delete_publisher(response_publisher_);
+      if (nullptr != impl::check_delete_publisher(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_publisher(status));
+      }
+    }
+    if (request_datareader_) {
+      // Assumption: subscriber is not null at this point.
+      status = request_subscriber_->delete_datareader(request_datareader_);
+      if (nullptr != impl::check_delete_datareader(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_datareader(status));
+      }
+    }
+    if (request_subscriber_) {
+      status = participant_->delete_subscriber(request_subscriber_);
+      if (nullptr != impl::check_delete_subscriber(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_subscriber(status));
+      }
+    }
+    if (request_topic_) {
+      status = participant_->delete_topic(request_topic_);
+      if (nullptr != impl::check_delete_topic(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_topic(status));
+      }
+    }
```

Comment:
OK, since `deinit` gets called from `destroy_requester__<srv_name>` and `destroy_responder__<srv_name>` I will change it to return an error string and cause those functions to fail if the string is not null.

also, this is a total bikeshed note, but I'm going to change the name from `deinit` to `teardown` or something else that's more specific.